### PR TITLE
Newsletters: add 314 (2024-08-02)

### PR DESCRIPTION
**Pull request by harding**
Happy π newsletter!

- [ ] Lede, releases/RCs, topic entries @harding 
- [ ] Merge summaries @Gustavojfe

### CODE REVIEW

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  smaller than its original chunk and it may sometimes result in several
+  sub-chunks.  Wuille demonstrates that the number of chunks and
+  sub-chunks is equal to the number of transactions, with each
+  transaction belonging to a unique chunk or sub-chunk.  That makes it
+  possible to precompute each transaction's chunk or sub-chunk, called
+  its _absorption set_, and associate that with the transaction.  Wuille
+  shows how the existing chunking algorithm already calculates each
+  transaction's absorption set.
+
+  When a miner has filled a template will all of the full chunks
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:65]:
```suggestion
  When a miner has filled a template with all of the full chunks
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  possible, it can take the precomputed absorption sets for all
+  transactions not yet included in the block and put them in feerate
+  order.  This only requires a single sort operation on a list with the
+  same number of elements as there are transactions in the mempool
+  (almost always less than a million with current defaults).  The best
+  feerate absorption sets (chunks and sub-chunks) can then be used them
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:71]:
```suggestion
  feerate absorption sets (chunks and sub-chunks) can then be used to
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  transactions from a cluster that have been included so far and
+  skipping any sub-chunks that don't fit or which have already had some
+  of their transactions included.
+
+  However, although linearization puts a list of chunks in the best
+  order for block inclusion, the individual transactions within a chunk
+  or sub-chunk are not guaranteed to be in the best order for only
+  including some of those transactions.  That can lead to non-optimal
+  selection when a block is nearly full.  For example, when only 300
+  vbytes remains, the algorithm might select a 200-vbyte transaction at
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:82]:
```suggestion
  vbytes remain, the algorithm might select a 200-vbyte transaction at
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  5 sats/vbyte (1,000 sats total) instead of two 150-vbyte transactions
+  at 4 sats/vbyte (1,200 sats total).
+
+  Wuille describes how precomputed absorption sets are especially useful
+  in this case: because they only require tracking the number of
+  transactions from each cluster that have been included so far, they
+  make it easy to restore to an earlier state in the template-filling
+  algorithm and replace the previously made choice with an alternative
+  choice to see if it results in collecting more total fees.  This
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:91]:
```suggestion
  to see if it results in collecting more total fees.  This
```

**Review by bitschmidty — COMMENTED** [review]
Looks good so far

**Comment by harding** [comment]
Edits made for @sipa comments (thanks!) and added summaries of the vulnerability disclosures.  Note: disclosure summary has a few FIXMEs for exact links and is in a separate commit to allow a clean revert in case they aren't published by the time we're scheduled to publish the newsletter.

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  transactions from each cluster that have been included so far, they
+  make it easy to restore to an earlier state in the template-filling
+  algorithm and replace the previously made choice with an alternative
+  to see if it results in collecting more total fees.  This
+  allows implementing a [branch-and-bound][] search that can try many
+  combinations of filling the last bit of block space in the hopes of
+  finding a better result than the simple algorithm.
+
+- **Hyperion network event simulator for the Bitcoin P2P network:**
+  Sergi Delgado [posted][delgado hyperion] to Delving Bitcoin about a
```
Comment (vostrnad) [inline _posts/en/newsletters/2024-08-02-newsletter.md:133]:
```suggestion
  Sergi Delgado [posted][delgado hyperion] to Delving Bitcoin about
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+  transaction announcements (`inv` inventory messages) to the proposed
+  [Erlay][topic erlay] method.
+
+## Releases and release candidates
+
+*New releases and release candidates for popular Bitcoin infrastructure
+projects.  Please consider upgrading to new releases or helping to test
+release candidates.*
+
+- [BDK 1.0.0-beta.1][] is a release candidate for "the beta first version of
```
Comment (vostrnad) [inline _posts/en/newsletters/2024-08-02-newsletter.md:146]:
```suggestion
- [BDK 1.0.0-beta.1][] is a release candidate for "the first beta version of
```
Also, I don't think this is even a "release candidate", it's certainly not called that in the release notes.

**Comment by Gustavojfe** [comment]
I just pushed a commit with all the merge summaries for this week's newsletter, left a comment on the one before the last one.

File: _posts/en/newsletters/2024-08-02-newsletter.md (hunk 6 above)
Comment (harding) [reply _posts/en/newsletters/2024-08-02-newsletter.md:146]: Ping BDK maintainer @notmandatory

Interesting.  I may have been misled by GitHub's UI calling it a "Pre-Release":



Comparing to LND, which uses `-beta` for its main releases, it looks like "Pre-Release" is set for their RCs but not their main releases:



latest

I suggest I make the transposition fix now but that we leave it described as an RC for this newsletter, see what @notmandatory says, and make any required changes to our announcement for next week's newsletter.

**Comment by harding** [comment]
@Gustavojfe 

> left a comment on the one before the last one.

Good question! "I'm not sure if I understand correctly if this is just an implementation bug or if there is actually a bug in the bolt12 specification."

It looks like the relevant behavior is correctly described in BOLT4, "The final recipient must verify that the blinded route is used in the right context [...] Note that it's important to use private information in that case, that senders cannot have access to."

So this seems like a case where LDK didn't do that and is now correcting it, although it was non-intuitive to me, so I'm glad were mentioning it where other LN implementations might notice it.  I'll update your excellent description of the solution to also briefly describe the attack to give it some extra emphasis.  Thanks!

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  - [Remote crash by sending excessive `addr` messages][]: before
+    Bitcoin 22.0 (released September 2021), a node that was told about
+    more than 2<sup>32</sup> other possible nodes would crash due to
+    exhaustion of a 32-bit counter.  This could be accomplished by an
+    attacker sending a large number of P2P `addr` messages (at least 4
+    million messages). <!-- assuming 1,000 addresses per addr message -->
+    Eugene Siegel [responsibly disclosed][topic responsible disclosures]
+    the vulnerability and a fix was included in Bitcoin Core 22.0.  See
+    [Newsletter #159][news159 bcc22387] for our summary of the fix,
+    which was written without us being aware that it patched a
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:37]:
```suggestion
    which was written without us knowing it patched a
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    vulnerability.
+
+  - [Remote crash on local network when UPnP enabled][]: before Bitcoin
+    22.0, nodes that enabled [UPnP][] for automatically configuring [NAT
+    traversal][] (disabled by default due to previous vulnerabilities,
+    see [Newsletter #310][news310 miniupnpc]) were vulnerable to a
+    malicious device on the local network repeatedly sending a variants
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:44]:
```suggestion
    malicious device on the local network repeatedly sending variants
```
or "a variant"

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+  1. If any transactions within the chunk depend on other unconfirmed
+     transactions, those other transactions must either be a part of
+     that chunk or appear in a chunk earlier in the ordered list of
+     chunks.
+
+  2. Each chunk must have an equal or higher feerate than the chunks
+     that come after it in the ordered list.
+
+  This allows every chunk from every cluster in the mempool to be placed
+  into a single list that is ordered from highest feerate to lowest
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:72]:
```suggestion
  into a single list ordered from highest feerate to lowest
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+  To work around this edge case problem, Wuille describes how large
+  chunks can be broken down into smaller _sub-chunks_ that can
+  considered for inclusion in the remaining block space based on their
+  feerates.  A sub-chunk can be created by simply removing the last
+  transaction in any existing chunk or sub-chunk that has two or more
+  transactions.  This will always produce at least one sub-chunk that is
+  smaller than its original chunk and it may sometimes result in several
+  sub-chunks.  Wuille demonstrates that the number of chunks and
+  sub-chunks is equal to the number of transactions, with each
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:100]:
```suggestion
  sub-chunks equals the number of transactions, with each
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+  command from conservative to economical. This change is based on user and
+  developer observations that the conservative mode often leads to overpayment
+  of transaction fees because it is less responsive to short-term fee market
+  drops than the economical mode when [estimating fees][topic fee estimation].
+
+- [Bitcoin Core #30408][] replaces the use of the wording "public key script" to
+  "output script" to refer to a `scriptPubKey` in the help text for the following
+  RPC commands `decodepsbt`, `decoderawtransaction`, `decodescript`, `getblock`
+  (if verbosity=3), `getrawtransaction` (if verbosity=2,3),  and `gettxout`.
+  This is the same wording that is used in the proposed BIP for transaction
```
Comment (LarryRuane) [inline _posts/en/newsletters/2024-08-02-newsletter.md:196]:
```suggestion
  This is the same wording used in the proposed BIP for transaction
```

**Review by LarryRuane — COMMENTED** [review]
> Happy π newsletter!

Not exactly. A few more small suggestions (I can't stay away!)

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    exhaustion of a 32-bit counter.  This could be accomplished by an
+    attacker sending a large number of P2P `addr` messages (at least 4
+    million messages). <!-- assuming 1,000 addresses per addr message -->
+    Eugene Siegel [responsibly disclosed][topic responsible disclosures]
+    the vulnerability and a fix was included in Bitcoin Core 22.0.  See
+    [Newsletter #159][news159 bcc22387] for our summary of the fix,
+    which was written without us being aware that it patched a
+    vulnerability.
+
+  - [Remote crash on local network when UPnP enabled][]: before Bitcoin
```
Comment (azuchi) [inline _posts/en/newsletters/2024-08-02-newsletter.md:40]:
```suggestion
  - [Remote crash on local network when UPnP enabled][]: before Bitcoin Core
```

File: _posts/en/newsletters/2024-08-02-newsletter.md [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+  Niklas Gögge [posted][goegge disclosure] to the Bitcoin-Dev mailing
+  list a link to announcements of
+  two vulnerabilities affecting versions of Bitcoin Core that have been
+  past their end of life since at least October 2022.  This follows a
+  previous disclosure last month of older vulnerabilities (see
+  [Newsletter #310][news310 disclosure]).  We summarize the disclosures
+  below:
+
+  - [Remote crash by sending excessive `addr` messages][]: before
+    Bitcoin 22.0 (released September 2021), a node that was told about
```
Comment (azuchi) [inline _posts/en/newsletters/2024-08-02-newsletter.md:29]:
```suggestion
    Bitcoin Core 22.0 (released September 2021), a node that was told about
```

**Comment by bitschmidty** [comment]
cherry-picked 

8429c99e56f62234255f0986c7c46f5f39918980
and 
7910e5781e6b1dc1b61c9c1479e343fae49392e4
from
https://github.com/bitcoinops/bitcoinops.github.io/tree/2024-08-02-newsletter as I believe they were pushed there in error.

**Comment by harding** [comment]
> cherry-picked
> 
> [8429c99](https://github.com/bitcoinops/bitcoinops.github.io/commit/8429c99e56f62234255f0986c7c46f5f39918980) and [7910e57](https://github.com/bitcoinops/bitcoinops.github.io/commit/7910e5781e6b1dc1b61c9c1479e343fae49392e4) from https://github.com/bitcoinops/bitcoinops.github.io/tree/2024-08-02-newsletter as I believe they were pushed there in error.

They were, sorry and thanks!

**Comment by bitschmidty** [comment]
Squashed and merged, thank you @harding @Gustavojfe @sipa @LarryRuane @azuchi @vostrnad for contributing and reviewing this week

File: _posts/en/newsletters/2024-08-02-newsletter.md (hunk 6 above)
Comment (notmandatory) [reply _posts/en/newsletters/2024-08-02-newsletter.md]: Hey sorry I just saw this in my notifications. BDK 1.0-beta.1 is the first release candidate for an upcoming final BDK 1.0 release. It's also worth noting that the original `bdk` rust crate has been renamed to `bdk_wallet` (https://crates.io/crates/bdk_wallet).

The old `bdk` crate used to include everything but was renamed `bdk_wallet` with lower layer modules extracted into their own crates (`bdk_chain`, `bdk_electrum`, `bdk_esplora`, `bdk_bitcoind_rpc`). Most users will only see `bdk_wallet` and which ever chain syncing crate(s) they want to use.

The purpose of the 1.0-beta.x releases is to:

1. help users update their apps to a ~stable 1.0 `bdk_wallet` API
2. improve testing and API docs
3. update and add examples and tutorials
4. update language bindings (kotlin, swift, python)

Hope this all makes sense, if not just let me know.

**Pull request merged**